src/core/eggmath.rkt

```diff
@@ -0,0 +1,213 @@
+    [(hash-has-key? herbie->egg-dict expr)
+     (symbol->string (hash-ref herbie->egg-dict expr))]
+    [else
+     (define new-key (string-append "h" (number->string (hash-count herbie->egg-dict))))
+     (define new-key-symbol (string->symbol new-key))
+         
+     (hash-set! herbie->egg-dict
+                expr
+                new-key-symbol)
+     (hash-set! egg->herbie-dict
+                new-key-symbol
+                expr)
+      new-key]))
+
+
+(define-struct (egg-add-exn exn:fail:user) ())
+
+;; result function is a function that takes the ids of the nodes
+;; egraph-add-exprs returns the result of result-function
+(define (egraph-add-exprs eg-data exprs result-function)
+  (define egg-exprs
+    (for/list ([expr exprs])
+      (expr->egg-expr expr eg-data)))
+  (debug #:from 'simplify (format "Sending expressions to egg_math:\n ~a"
+                                  (string-join (map ~a egg-exprs) "\n ")))
+      
+  (define expr-results
+    (for/list ([expr egg-exprs])
+      (egraph_add_expr (egraph-data-egraph-pointer eg-data) expr)))
+  
+  (define node-ids
+    (for/list ([result expr-results])
+      (if (EGraphAddResult-successp result)
+          (EGraphAddResult-id result)
+          (raise (egg-add-exn
+               (string-append "Failed to add expr to egraph")
+               (current-continuation-marks))))))
+
+  (define res (result-function node-ids))
+
+  (for/list ([result expr-results])
+    (egraph_addresult_destroy result))
+
+  res)
+
+
+(module+ test
+  (require (submod "../syntax/rules.rkt" internals))
+  
+  (define all-simplify-rules
+    (for/append ([rec (*rulesets*)])
+      (match-define (list rules groups _) rec)
+      (if 
+       (set-member? groups 'simplify)
+       rules
+       '())))
+
+  ;; check that no rules in simplify match on bare variables
+  ;; this would be bad because we don't want to match type-specific operators on a value of a different type
+  (for ([rule all-simplify-rules])
+    (check-true
+     (or
+      (not (symbol? (rule-input rule)))
+      (set-member? fpconstants (rule-input rule)))
+     (string-append "Rule failed: " (symbol->string (rule-name rule)))))
+
+  
+  (check-equal?
+   (egraph-run
+    (lambda (egg-graph)
+      (egraph-add-exprs
+       egg-graph
+       (list '(+ x 2) '(+ x 1) '(+ x 0) '(+ x 23) 0 1)
+       (lambda (node-ids)
+         (egraph-run-rules egg-graph 100 (*simplify-rules*) true)
+         (for/list [(node-id node-ids)]
+           (egg-expr->expr (egraph-get-simplest egg-graph node-id) egg-graph))))))
+   (list '(+ x 2) '(+ x 1) 'x '(+ x 23) 0 1))
+
+  (define expr-list
+    (list
+     '(+ x y)
+     '(+ y x)
+     '(- 2 (+ x y))
+     '(- z (+ (+ y 2) x))))
+  
+  (check-equal?
+   (egraph-run
+    (lambda (egg-graph)
+      (for/list ([expr
+                  expr-list])
+        (expr->egg-expr expr egg-graph))))
+   (list
+    "(+ h1 h0)"
+    "(+ h0 h1)"
+    "(- 2 (+ h1 h0))"
+    "(- h2 (+ (+ h0 2) h1))"))
```

Comment:
Good test, but I'd combine the input and output lists into a hash table. Potentially also make a `test-suite` and call `check-equal?` in a `for` loop, so that you'll know which expression failed if the test fails.